### DISCUSSION TITLE
Deploy Next.js with Passenger

### DISCUSSION THREAD

**Discussion by wengair** [vercel/next.js#17289] [Help] [answered]
Hi!
I'm trying to deploy my first Next.js React app with Passenger.
Passenger is required and installed by my server host.

Since Passenger is looking for `app.js` to start the application, I'd like to know how can I connect Next.js with Passenger in `app.js`?
Shoud I put all my codes in another folder and require it in `app.js`?
Or I can just create it in my root folder?

Any help would be appreciated!

**Comment by jamesmosier** [comment]
Could you create an `app.js` file and then put in it:

```js
const { execSync } = require('child_process');
const stdout = execSync('npm start'); // or yarn start
```

**Reply by wengair** [reply]
Thank you for the reply!
I tried your method by adding `app.js` in my project's root folder, but still got the same error page.
![img](https://imgur.com/YaqhxXJ.png)
I also got the same error page when I tried something else in my `app.js`, so I guess Passenger does monitor `app.js` but not 100% sure.

**Comment by iaurg** [comment]
Hi @wengair 

I needed to deploy into an Apache server with Passenger one time and the solution I've found is to use a server.js (or any name) file to start the server, then configure passenger to start using that file:

Start command into package.json: `"start": "NODE_ENV=production node server.js"`

Entry file:
```js
const express = require('express');
const next = require('next');

const { join } = require('path');
const { parse } = require('url');

const port = parseInt(process.env.PORT, 10) || 3000;
const dev = process.env.NODE_ENV !== 'production';
const app = next({ dev });
const handle = app.getRequestHandler();

app.prepare().then(() => {
  const server = express();

  redirects.forEach(({ from, to, type = 301, method = 'get' }) => {
    server[method](from, (req, res) => {
      res.redirect(type, to);
    });
  });

  server.get('/static', (req, res) =>
    ssrCache({ req, res, pagePath: '/static' })
  );

  server.get('*', (req, res) => handle(req, res));

  server.listen(port, err => {
    if (err) throw err;
    console.log(`> Ready on http://localhost:${port}`);
  });
});
```

I think your Next app needs to start into a custom server

You cand find some tips into Vercel docs about [custom server](https://nextjs.org/docs/advanced-features/custom-server)

I expect it can help you.

**Reply by wengair** [reply]
Hi
Since that host can't change the entry file of Passenger (at least not in the panel, I'm still waiting for their engineer to reply because passenger command is not available on their server), I tried to combine yours and @jamesmosier's method but still can't get it to work.
Is there anyway to run `app.js` as if run this `server.js`?

Also, just want to make sure that no matter what method I use, I still need to `npm run build`, is that correct?

**Reply by jamesmosier** [reply]
Yes you would still need to do a `npm run build` before uploading the files to your hosting provider.

**Comment by wengair** [comment, marked as answer]
I found another way to do this.
https://0x7f.dev/2020/03/31/next.js-with-passenger-and-nginx/
I just copy and paste the code from the link above into `app.js`
```
const path = require('path');

const nextPath = path.join(__dirname, 'node_modules', '.bin', 'next');

process.argv.length = 1;
process.argv.push(nextPath, 'start');

require(nextPath);
```
That's all, it works!

The only thing worth noticed is that either I have to put the `app.js` in my project's root folder.
Or I could put my project in another folder, but then I have to copy `public`, `node_module`, `.next`, API routes, and maybe more files to the same folder as `app.js`. Otherwise, the website can run but can't access lots of resources.

**Reply by gremo** [reply]
Can't make it work. Are you sure it works with the latest version of Next?

**Reply by rchrd2** [reply]
I got this to work on Dreamhost shared hosing by adding a port. Eg
```
const path = require('path');
const nextPath = path.join(__dirname, 'node_modules', '.bin', 'next');
process.argv.length = 1;
process.argv.push(nextPath, 'start');
process.env.NODE_ENV = 'production';
process.env.PORT = 8888
require(nextPath);
```

**Reply by ghost** [reply]
just to bump an old post, this way still works with plesk, nginx, passenger and next 12.1 in 2022

**Reply by Arctomachine** [reply]
How do you run build script when working this way?

**Comment by Maxime-J** [comment]


**Reply by Arctomachine** [reply]
Is there a reason for this code to start 2 servers?
```
const start = require('next/dist/cli/next-start')
start.nextStart(['-p', process.env.PORT || 3000, '-H', process.env.HOSTNAME || '127.0.0.1'])
```
Outputs
```
ready - started server on 127.0.0.1:3000, url: http://127.0.0.1:3000
ready - started server on 0.0.0.0:3000, url: http://localhost:3000
Error: Could not find a production build
```
And opening index page then returns 404 and
```
Unhandled Runtime Error
TypeError: anchorElement is null
```
Running `next build` prior to this script does not help.
Running `next start` manually starts normal server as usually.

**Reply by Maxime-J** [reply]
Weird, these two lines alone start one server, and shouldn't have problems with production build.

Does it produce the same output If you only put these lines in a script and call it from the same path you call `next start`?
`node twoLinesScript.js` instead of `next start`

**Reply by Arctomachine** [reply]
This is exactly how I do it.
I tried to remove all arguments from function call, so this is what happens. After running this file for first time I get
```
ready - started server on 0.0.0.0:3000, url: http://localhost:3000
ready - started server on 0.0.0.0:3000, url: http://localhost:3000
wait  - compiling...
event - compiled client and server successfully in 618 ms (135 modules)
```
Then, when I open localhost:3000 in browser, I get
```
warn  - You have added a custom /_error page without a custom /404 page. This prevents the 404 page from being auto statically optimized.
See here for info: https://nextjs.org/docs/messages/custom-error-no-custom-404
```
And next time server fails to start because it somehow invalidates built app
```
Error: Could not find a production build in
```
The error at 404 page is this
```
Unhandled Runtime Error

TypeError: anchorElement is null
Call Stack
options.insert
webpack-internal:///./styles/globals.scss (16:34)
```
I do have such a file in my project, and I have `sass` package.
Strange thing 1. Even if I convert .scss to .css and remove unsupported code, the error trace keeps saying about .scss file.
Strange thing 2. There is no such coordinates in this file. The last symbol of line 16 is 16:20. Here are lines 15-17 from this file.
```
  font-style: normal;
  font-weight: 300;
}
```
Any idea what can be wrong with my project?

**Reply by Maxime-J** [reply]
Strange, I'm curious, do you mind sharing a repo of your project ?
We could continue this conversation there.

**Reply by Arctomachine** [reply]
I think I found where second server is coming from. Before we start looking into project itself, can you confirm if this is intended?
Having only import line `const start = require('next/dist/cli/next-start')` in the file already starts a server. But also `const build = require('next/dist/cli/next-build')` starts server as well. By the look of console, it seems to start dev server. Contents of  .next folder also look similar to what appears after running `next dev`.
```
ready - started server on 0.0.0.0:3000, url: http://localhost:3000
```

**Reply by Maxime-J** [reply]
Precisely not, `const start = require('next/dist/cli/next-start')` alone in a script does nothing.
If it does on your end, there must be something else, did you try?

[https://github.com/vercel/next.js/blob/canary/packages/next/cli/next-start.ts](https://github.com/vercel/next.js/blob/canary/packages/next/cli/next-start.ts) only export `nextStart`.
And it's what is used internally with `next start`

Perhaps it tries to respawn an already launched server ??

It would also be interesting to see if you have that behaviour with that way https://github.com/vercel/next.js/discussions/17289#discussioncomment-85258

**Reply by Arctomachine** [reply]
I tried closing all instances of command lines just in case. But it would start at port 3001 anyway if there was server already running.
The other solution is not working for me. Running target script directly from command line works, but running it from inside .js file just like instructed gives error
```
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
          ^^^^^^^

SyntaxError: missing ) after argument list
```

**Reply by Maxime-J** [reply]
Hi @Arctomachine , I was using Next 12.1.0, and I'm facing now the same issue as yours with Next 12.2.0.
I will delete this outdated solution soon.

**Comment by JorgeReyCristino** [comment]
Hi!

Does anybody know how to do it with Next 12.2?
I would like to avoid using a custom server to deploy my app in Passenger.

Thanks!

**Comment by zafon18** [comment]
I came across such a solution. When changing the port to 3000. The application starts

```
const { createServer } = require("http");
var https = require('https');

const { parse } = require("url");
const next = require("next");
const dev = process.env.NODE_ENV !== "production";

const fs = require("fs");


const port = !dev ? **process.env.PORT** : 8000;

// Create the Express-Next App
const app = next({ dev });
const handle = app.getRequestHandler();


app
  .prepare()
  .then(() => {
    createServer((req, res) => {
      const parsedUrl = parse(req.url, true);
      const { pathname, query } = parsedUrl;
      handle(req, res, parsedUrl);
      console.log("App.js> pathname", pathname);
    }).listen(port, (err) => {
      if (err) throw err;

      console.log("process.env.PORT:"+process.env.PORT);
      console.log(`APP.js> Ready on http://localhost:${port}`);

    });
  })
  .catch((ex) => {
    console.error(ex.stack);
    process.exit(1);
  });
```
source:
https://stackoverflow.com/posts/76450645/revisions
